Generate parameter names

When an interface definition leaves its method parameters anonymous, it would be nice if `impl` would generate a name for the variable.

This is granted non-trivial, since there are scoping problems - named results, the receiver, and other parameters might all collide.

Or at least error with a message that allows the user to understand why it crashed. Currently it looks like this:

```
panic: 13:30: expected operand, found ',' (and 1 more errors)

goroutine 1 [running]:
main.genStubs(0x7ffeefbff5f1, 0x16, 0xc420189200, 0x3, 0x4, 0x0, 0x0, 0x11b50c0)
	/Users/zwass/dev/go/src/github.com/groob/mockimpl/impl.go:294 +0x340
main.main()
	/Users/zwass/dev/go/src/github.com/groob/mockimpl/impl.go:373 +0x19e
datastore.go:12: running "mockimpl": exit status 2
```

--- a/impl.go
+++ b/impl.go
@@ -215,7 +215,14 @@ func (p Pkg) funcsig(f *ast.Field, cmap ast.CommentMap) Func {
 	typ := f.Type.(*ast.FuncType)
 	if typ.Params != nil {
 		for _, field := range typ.Params.List {
-			fn.Params = append(fn.Params, p.params(field)...)
+			for _, param := range p.params(field) {
+				// only for method parameters:
+				// assign a blank identifier "_" to an anonymous parameter
+				if param.Name == "" {
+					param.Name = "_"
+				}
+				fn.Params = append(fn.Params, param)
+			}
 		}
 	}
 	if typ.Results != nil {

--- a/impl_test.go
+++ b/impl_test.go
@@ -115,7 +115,7 @@ func TestFuncs(t *testing.T) {
 				},
 				{
 					Name:   "Write",
-					Params: []Param{{Type: "[]byte"}},
+					Params: []Param{{Name: "_", Type: "[]byte"}},
 					Res:    []Param{{Type: "int"}, {Type: "error"}},
 				},
 				{
@@ -128,8 +128,11 @@ func TestFuncs(t *testing.T) {
 			iface: "http.Handler",
 			want: []Func{
 				{
-					Name:   "ServeHTTP",
-					Params: []Param{{Type: "http.ResponseWriter"}, {Type: "*http.Request"}},
+					Name: "ServeHTTP",
+					Params: []Param{
+						{Name: "_", Type: "http.ResponseWriter"},
+						{Name: "_", Type: "*http.Request"},
+					},
 				},
 			},
 		},
@@ -426,6 +429,80 @@ func TestValidMethodComments(t *testing.T) {
 				},
 			},
 		},
+		{
+			iface: "github.com/josharian/impl/testdata.Interface3",
+			want: []Func{
+				Func{
+					Name: "Method1",
+					Params: []Param{
+						Param{
+							Name: "_",
+							Type: "string",
+						}, Param{
+							Name: "_",
+							Type: "string",
+						}},
+					Res: []Param{
+						Param{
+							Name: "",
+							Type: "string",
+						},
+						Param{
+							Name: "",
+							Type: "error",
+						},
+					}, Comments: "// Method1 is the first method of Interface3.\n",
+				},
+				Func{
+					Name: "Method2",
+					Params: []Param{
+						Param{
+							Name: "_",
+							Type: "int",
+						},
+						Param{
+							Name: "arg2",
+							Type: "int",
+						},
+					},
+					Res: []Param{
+						Param{
+							Name: "_",
+							Type: "int",
+						},
+						Param{
+							Name: "err",
+							Type: "error",
+						},
+					},
+					Comments: "// Method2 is the second method of Interface3.\n",
+				},
+				Func{
+					Name: "Method3",
+					Params: []Param{
+						Param{
+							Name: "arg1",
+							Type: "bool",
+						},
+						Param{
+							Name: "arg2",
+							Type: "bool",
+						},
+					},
+					Res: []Param{
+						Param{
+							Name: "result1",
+							Type: "bool",
+						},
+						Param{
+							Name: "result2",
+							Type: "bool",
+						},
+					},
+					Comments: "// Method3 is the third method of Interface3.\n",
+				},
+			},
+		},
 	}
 
 	for _, tt := range cases {
@@ -451,6 +528,10 @@ func TestStubGeneration(t *testing.T) {
 		{
 			iface: "github.com/josharian/impl/testdata.Interface2",
 			want:  testdata.Interface2Output,
+		},
+		{
+			iface: "github.com/josharian/impl/testdata.Interface3",
+			want:  testdata.Interface3Output,
 		},
 	}
 	for _, tt := range cases {

--- a/testdata/interfaces.go
+++ b/testdata/interfaces.go
@@ -28,6 +28,26 @@ type Interface2 interface {
 	Method3(arg1 interface{}, arg2 interface{}) (result interface{}, err error)
 }
 
+// Interface3 is a dummy interface to test the program output. This interface
+// tests generation of method parameters and results.
+//
+// The first method tests the generation of anonymous method paramaters and
+// results.
+//
+// The second method tests the generation of method parameters and results where
+// the blank identifier "_" is already used in the names.
+//
+// The third method tests the generation of method parameters and results that
+// are grouped by type.
+type Interface3 interface {
+	// Method1 is the first method of Interface3.
+	Method1(string, string) (string, error)
+	// Method2 is the second method of Interface3.
+	Method2(_ int, arg2 int) (_ int, err error)
+	// Method3 is the third method of Interface3.
+	Method3(arg1, arg2 bool) (result1, result2 bool)
+}
+
 // Interface1Output is the expected output generated from reflecting on
 // Interface1, provided that the receiver is equal to 'r *Receiver'.
 var Interface1Output = `// Method1 is the first method of Interface1.
@@ -67,6 +87,25 @@ func (r *Receiver) Method2(arg1 float64, arg2 float64) (result float64, err erro
 	Method3 is the third method of Interface2.
 */
 func (r *Receiver) Method3(arg1 interface{}, arg2 interface{}) (result interface{}, err error) {
+	panic("not implemented")
+}
+
+`
+
+// Interface3Output is the expected output generated from reflecting on
+// Interface3, provided that the receiver is equal to 'r *Receiver'.
+var Interface3Output = `// Method1 is the first method of Interface3.
+func (r *Receiver) Method1(_ string, _ string) (string, error) {
+	panic("not implemented")
+}
+
+// Method2 is the second method of Interface3.
+func (r *Receiver) Method2(_ int, arg2 int) (_ int, err error) {
+	panic("not implemented")
+}
+
+// Method3 is the third method of Interface3.
+func (r *Receiver) Method3(arg1 bool, arg2 bool) (result1 bool, result2 bool) {
 	panic("not implemented")
 }
 

